File: 127-word-ladder/word-ladder.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        
        Set<String> hs= new HashSet<>(wordList);
        if(beginWord.equals(endWord)) return 1;
        Queue<String> q= new LinkedList<>();
        q.offer(beginWord);
        int level=1;
        while(!q.isEmpty()){
            int size=q.size();
            for(int length=0; length< size ; length++){
                String word= q.poll();;
                for(int i=0 ; i< word.length(); i++){
                char[] chArray= word.toCharArray();
                    for(char j='a'; j<='z'; j++){
                        if(j==chArray[i]) continue;
                        chArray[i]=j;
                        String newWord= new String(chArray);
                        if(hs.contains(newWord)){
                            if(newWord.equals(endWord)) return level+1;
                            System.out.println(newWord);
                            q.offer(newWord);
                            hs.remove(newWord);
                             }
                        }
        
                    }
             
            }
            level++;
            
        }
        return 0;
    }
}

File: 127-word-ladder/word-ladder.py
class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        wordset= set(wordList)
        if endWord not in wordset:
             return 0
        if beginWord==endWord:
             return 1

        q=deque()
        q.append(beginWord)
        steps=1

        while q:
            
            for _ in range(len(q)):
                curr= q.popleft()
                if curr==endWord:
                     return steps
                chArray= list(curr)
               

                for i,ch in enumerate(chArray):
                    original =chArray[i]
                    for letter in range(ord('a'), ord('z')+1):
                        chArray[i]=chr(letter)
                        if letter== original:
                            continue
                        currString= "".join(chArray)
                        if currString in wordset:
                            wordset.remove(currString)
                            q.append(currString)
                    chArray[i]=original
                
            steps+=1
        return 0
                        


